### ISSUE TITLE
GHC HEAD error

### ISSUE THREAD

**Issue by ggreif** [ekmett/lens#667]
```
src/Control/Lens/Internal/Indexed.hs,
dist/build/Control/Lens/Internal/Indexed.o )

src/Control/Lens/Internal/Indexed.hs:238:10: error:
    ? Couldn't match type ?Indexed i? with ?(->)?
      Inaccessible code in
        a type expected by the context:
          Indexed i ~ (->) => q (a -> b) r
    ? In the expression:
        Control.Lens.Internal.Indexed.$dmconjoined @Indexed i
      In an equation for ?conjoined?:
          conjoined = Control.Lens.Internal.Indexed.$dmconjoined @Indexed i
      In the instance declaration for ?Conjoined (Indexed i)?
    ? Relevant bindings include
        conjoined :: (Indexed i ~ (->) => q (a -> b) r)
                     -> q (Indexed i a b) r -> q (Indexed i a b) r
          (bound at src/Control/Lens/Internal/Indexed.hs:238:10)
cabal: Leaving directory '/home/ggreif/.UCSITmpDir/cabal-tmp-25477/lens-4.14'
cabal: Error: some packages failed to install:
clash-prelude-0.11 depends on ghc-typelits-natnormalise-0.4.6 which failed to
install.
ghc-typelits-natnormalise-0.4.6 failed during the building phase. The
exception was:
ExitFailure 1
lens-4.14 failed during the building phase. The exception was:
ExitFailure 1
```

**Comment by RyanGlScott** [comment]
I've opened [GHC Trac #12466](https://ghc.haskell.org/trac/ghc/ticket/12466) for this issue.

**Comment by RyanGlScott** [comment]
A workaround for this issue is to manually implement `conjoined` for `Conjoined` instances. That is, apply this change:

```diff
diff --git a/src/Control/Lens/Internal/Indexed.hs b/src/Control/Lens/Internal/Indexed.hs
index e6d042e..46574da 100644
--- a/src/Control/Lens/Internal/Indexed.hs
+++ b/src/Control/Lens/Internal/Indexed.hs
@@ -235,6 +235,9 @@ instance Conjoined (Indexed i) where
   distrib (Indexed iab) = Indexed $ \i fa -> iab i <$> fa
   {-# INLINE distrib #-}
 
+  conjoined _ r = r
+  {-# INLINE conjoined #-}
+
 instance i ~ j => Indexable i (Indexed j) where
   indexed = runIndexed
   {-# INLINE indexed #-}
diff --git a/src/Control/Lens/Reified.hs b/src/Control/Lens/Reified.hs
index c5c0dc8..d6e76c3 100644
--- a/src/Control/Lens/Reified.hs
+++ b/src/Control/Lens/Reified.hs
@@ -190,7 +190,9 @@ instance Representable ReifiedGetter where
 instance Costrong ReifiedGetter where
   unfirst l = Getter $ to $ unfirst $ view (runGetter l)
 
-instance Conjoined ReifiedGetter
+instance Conjoined ReifiedGetter where
+  conjoined _ r = r
+  {-# INLINE conjoined #-}
 
 instance Strong ReifiedGetter where
   first' l = Getter $ \f (s,c) ->
```

**Comment by RyanGlScott** [comment]
This particular issue has been fixed since [GHC Trac #12466](https://ghc.haskell.org/trac/ghc/ticket/12466) has been fixed, so the `lens` library should build on GHC HEAD now. The tests still fail on GHC HEAD, however, due to #678.
